File: Life.js
(function Life () {
  var LIVE = "XX ",
    DEAD = "   "

  function Cell (state, neighbours) {
    this.state = state;
    this.neighbours = neighbours;
  }

  Cell.prototype.getState = function () {
    return this.state;
  }

  Cell.prototype.live = function () {
    if (this.state == LIVE) {
      if (this.neighbours < 2) {
        this.state = DEAD;
      } else if (this.neighbours == 2 || this.neighbours == 3) {
        this.state = LIVE;
      } else if (this.neighbours > 3) {
        this.state = DEAD;
      }
    }

    if (this.state == DEAD) {
      if (this.neighbours == 3) {
        this.state = LIVE;
      }
    }
  }

  function Board (height, width, board) {
      this.height = height || 12;
      this.width = width || 38;
      this.board = board || [
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD],
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD],
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD],
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD],
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD],
        [DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD],
        [DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,LIVE,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD],
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD],
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD],
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD],
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD],
        [DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD]
      ];
  }

  Board.prototype.step = function (yPos, xPos) {
    var cell, neighbours,
      output = "",
      updatedBoard = [];

    for (y = 0; y < this.height; y++) {
      updatedBoard[y] = [];
      for (x = 0; x < this.width; x++) {
        neighbours = this.countNeighbours(y, x);
        cell = new Cell(this.board[y][x], neighbours);
        cell.live();
        output += cell.getState();
        updatedBoard[y].push(cell.getState());
      }
      output += "\n";
    }

    this.board = updatedBoard;
    console.log(output);
  }

  Board.prototype.isSamePosition = function (y1, x1, y2, x2) {
    return y1 == y2 && x1 == x2;
  }

  Board.prototype.countNeighbours = function (yPos, xPos) {
    var x, y, cell,
      count = 0,
      startY = yPos - 1 > 0 ? yPos - 1 : 0,
      startX = xPos - 1 > 0 ? xPos - 1 : 0,
      endY = yPos + 1 < this.height ? yPos + 1 : this.height - 1,
      endX = xPos + 1 < this.width ? xPos + 1 : this.width - 1;

    for (y = startY; y <= endY; y++) {
      for (x = startX; x <= endX; x++) {
        cell = new Cell(this.board[y][x], 0);
        if (cell.getState() == LIVE && !this.isSamePosition(y, x, yPos, xPos)) {
          count = count + 1;
        }
      }
    }

    return count;
  }

  function Game () {
    this.board = new Board();
  }

  Game.prototype.run = function () {
    setTimeout(function () {
      this.board.step();
      this.run()
    }.bind(this), 200);
  }

  game = new Game();
  game.run();
})();

File: Life.cpp
#include <iostream>
#include <string>
#include <unistd.h>

using namespace std;

namespace Life {
  const string LIVE = "XX ";
  const string DEAD = "   ";

  class CRectangle {
    private:
      int x, y;
    public:
      void set_values (int,int);
      int area () {return (x*y);}
      string test ();
  };

  void CRectangle::set_values (int a, int b) {
    x = a;
    y = b;
  }

  string CRectangle::test () {
    return LIVE;
  }

  class Cell {
    private:
      int neighbours;
      string state;
    public:
      Cell (string, int);
      string getState ();
      void live ();
  };

  Cell::Cell (string state, int neighbours) {
    this->state = state;
    this->neighbours = neighbours;
  }

  string Cell::getState () {
    return state;
  }

  void Cell::live () {
    if (state == LIVE) {
      if (neighbours < 2) {
        state = DEAD;
      } else if (neighbours == 2 || neighbours == 3) {
        state = LIVE;
      } else if (neighbours > 3) {
        state = DEAD;
      }
    }

    if (state == DEAD) {
      if (neighbours == 3) {
        state = LIVE;
      }
    }
  }

  class Board {
    private:
      int width;
      int height;
      string board[12][38];
    public:
      Board ();
      void step ();
      bool isSamePosition (int, int, int, int);
      int countNeighbours (int, int);
  };

  Board::Board () : board {
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD},
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD},
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD},
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD},
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD},
      {DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD},
      {DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,LIVE,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD},
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD},
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD},
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,LIVE,LIVE,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD},
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD},
      {DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD,DEAD}
  } {
    this->height = 12;
    this->width = 38;
  };

  void Board::step () {
    int neighbours;
    string output = "";
    string updatedBoard[12][38];

    for (int y = 0; y < height; y++) {
      for (int x = 0; x < width; x++) {
        neighbours = countNeighbours(y, x);
        Cell cell (board[y][x], neighbours);
        cell.live();
        output += cell.getState();
        updatedBoard[y][x] = cell.getState();
      }
      output += "\n";
    }

    for (int y = 0; y < height; y++) {
      for (int x = 0; x < width; x++) {
        board[y][x] = updatedBoard[y][x];
      }
    }

    cout << output;
  }

  bool Board::isSamePosition (int y1, int x1, int y2, int x2) {
    return y1 == y2 && x1 == x2;
  }

  int Board::countNeighbours (int yPos, int xPos) {
    int count = 0;

    int startY = yPos - 1 > 0 ? yPos - 1 : 0;
    int startX = xPos - 1 > 0 ? xPos - 1 : 0;
    int endY = yPos + 1 < height ? yPos + 1 : height - 1;
    int endX = xPos + 1 < width  ? xPos + 1 : width - 1;

    for (int y = startY; y <= endY; y++) {
      for (int x = startX; x <= endX; x++) {
        Cell cell (board[y][x], 0);
        if (cell.getState() == LIVE && !isSamePosition(y, x, yPos, xPos)) {
          count = count + 1;
        }
      }
    }

    return count;
  }

  class Game {
    private:
      Board board;
    public:
      Game ();
      void run ();
  };

  Game::Game () {
    Board board();
  }

  void Game::run () {
    while (true) {
      board.step();
      usleep(200000);
    }
  }
}

int main () {
  Life::Game game;
  game.run();
  return 0;
}
